Parse all command line options in repl.c

jakebolewski opened the pull request:
Addresses #9384.  Currently the options that are not in `jl_compileropts_t` are stored into a new `jl_options_t` struct.  This struct is then loaded when processing the command line options in `base/client.jl`.  This PR adds some sanity check tests for command line arguments as well.

I think it would be better to rename `jl_compileropts_t` and just have one option struct for all command line arguments (and api to query the arguments from Julia).

I've deprecated `--no-history-file` and `--no-startup` flags in favor of  `--history-file={yes|no}` and `--startup-file={yes|no}` so all command line options have a similar structure.

jakebolewski committed:
Parse all command line options in repl.c

 - consolidate all compiler / cmdline options into jl_options_t struct in julia.h

 - add options.jl to base/ with an immutable type JLOptions that
   reflects the jl_options_t struct

 - add --procs=<n> command line flag (equivalent to -p <n>)

 - add --history-file={yes|no} and --startup-file={yes|no} cmdline opts

 - --worker command line arguments changed to --worker,
   --worker=default, or --worker=custom

 - deprecate -f, -F, --no-startup, --no-history-file cmdline flags

 - add tests for cmdline arguments in test/cmdlineargs.jl

 - modify test/Makefile and tests to use long command line option,
   add command line argument tests to runtests.jl

 - update relevant docs in the manual and manpages

samoconnor commented:
Nice one.
It was confusing for me to try to find how the "-e" option was handled because it was not in repl.c.
See #9468 and specifically: https://github.com/samoconnor/julia/blame/doc_notes_branch/doc/devdocs/init.rst#L22

jakebolewski commented:
Rebased and squashed.  I consolidated all options into a single struct as talked about.  The old `jl_compileropts` struct was exported so I'm unsure if we consider this part of the "public" C API.  On the julia side, the option struct now lives in `base/options.jl`.  I also added more tests and updated the documentation.

tkelman commented on ui/repl.c:
```diff
@@ -44,62 +44,101 @@ static int imagepathspecified = 0;
 {
-    static char* shortopts = "+H:hJ:C:O";
+    enum { machinefile = 300,
+	       color,
```
no tabs whenever you rebase this

jakebolewski replied:
Ah, thanks.  I have to change my Vim defaults.

tkelman replied:
note that #9732 might get merged before this and touches the same code

timholy replied:
Sorry about that, @jakebolewski.

tkelman commented on test/cmdlineargs.jl:
```diff
@@ -0,0 +1,157 @@
+    @test readall(`$exename -v`) == readall(`$exename --version`)
+
+    # --help
+    @unix_only begin
```
fixed by #9748, shouldn't need the `@unix_only` when you rebase

jakebolewski replied:
Cool, thanks @tkelman

jakebolewski commented:
Ok, I rebased just rebased.  It needs some review as it touches a lot of code and I'm unsure if shoving everything  into one struct is the cleanest way to do this.

ihnorton commented on base/client.jl:
```diff
@@ -210,113 +208,122 @@ function init_bind_addr(args::Vector{UTF8String})
+    global process_options
+    function process_options(opts::JLOptions, args::Vector{UTF8String})
+        if !isempty(args) && (arg = first(args); arg[1] == '-' && in(arg, reqarg))
+            println(STDERR, "julia: option `$arg` is missing an argument")
```
couldn't we do this by checking the return value of getopt_long? http://man7.org/linux/man-pages/man3/getopt.3.html#RETURN_VALUE

jakebolewski replied:
That would be possible if the long option struct array was an exported global.

ihnorton commented:
aesthetics (ignore, more or less)
- `looks_like(deity_obj) && smells_like(deity_obj) == true`
- at very least, having something called `eval` in there feels weird

other than my one inline comment, the actual code LGTM. lots of repetition but i guess that is unavoidable here. it would be nice if we could store all of the text-based longopts in a dictionary, but i guess that would be harder to use (or not available at all when needed) from some parts of the C code.

ihnorton commented:
Bump

jakebolewski commented:
I'm waiting for @samoconnor's PR to get merged.

tkelman commented:
#9468, #9450, or both?

jakebolewski commented:
9450, he has been rebasing it too many times I don't want to add to his pain :-)

JeffBezanson commented:
9450 merged.

StefanKarpinski commented:
See also https://github.com/JuliaLang/julia/issues/10226 – argument parsing should stop at `--`.

jakebolewski commented:
The c library that we use for unix `opt` parsing, `ui/getopt.c` would have to be changed to support this.

StefanKarpinski commented:
How feasible is just moving all of the option parsing into Julia? Most of the REPL is in Julia these days so it seems like it shouldn't be too crazy to do that.

jakebolewski commented:
I don't think too feasible at the moment.  How do you turn off boundschecking for generated julia code that is being executed to parse the option to turn off boundschecking?

tkelman commented:
Moving things to Julia is the opposite direction from what this PR does. And there's a bootstrapping problem for the flags that control compiler options, which is a good chunk of them.

Note that we use the system version of getopt everywhere except in the (currently broken) MSVC build.

samoconnor commented:
@tkelman says that system getopt is used by default.
Double dash end of args should already be handled by the system getopt:

Linux: _The special argument "--" forces an end of option-scanning regardless of the scanning mode_
BSD: _The interpretation	of options in the argument list	may be cancelled by
     the option	`--' (double dash) which causes	getopt() to signal the end of
     argument processing and return -1._

Is there an overall policy for how cross-platform issue should be handled in Julia?

e.g. my usual rules for a non-GUI cross-platform system would be:
* Use POSIX-ish APIs
* Use GNU extensions, BSD extensions, etc wherever these have been adopted by the important platforms (i.e. Linux and OSX)
* Disallow #ifdef WIN32 in the main codebase
* Support windows by by linking to a compatibility library that provides whatever POSIX-ish functions are missing on windows (e.g. ui/getopt.c and there are lots of implementation of posix functions for windows out there...)
* Allow building against cygwin as an alternative.

The alternatives are:
* Code full of #ifdef WIN32
* Only supporting the smallest common subset of OS-librarty features
* Building yet-another-cross-platform-API-wrapper-layer-library-thingamabob.

tkelman commented:
@samoconnor everything you describe is what we use libuv for...

samoconnor commented:
True, libuv is used (and seems like a decent choice) for network and IO stuff, but there is still a lot of #ifdef _OS_WINDOWS_ elsewhere.

```
find julia/src/ -name '*.c' | xargs grep -l OS_WINDOWS | wc -l
26
```

tkelman commented:
Well going out of your way to have fewer ifdefs if you don't deem Windows an "important platform" is an option for some projects. Cygwin is not an option for Julia since Cygwin doesn't implement any asynchronous IO required by libuv.

For things that libuv doesn't cover, like a different set of headers being available, libc functions like `strtod` having different functionality, handling of paths and executables and shared libraries and exception handling and debug info and calling conventions (and and and) being different, etc, they're handled on a case-by-case basis. I don't think there's one unifying policy, other than things mostly work now, and we like making them work better where possible. Refactorings that clean things up are also welcome when they don't harm existing functionality.

samoconnor commented:
@tkelman ... that all sounds reasonable.
But, I think something being an "important platform" is an emotive way to think about it.
I think you can have windows be an "important platform" while simultaneously saying "libuv + POSIX(-ish) is the primary OS API layer, and windows specific implementations are hidden behind that". The former is a matter of marketing and customer support, the latter is a matter of architectural discipline and maintainability.

I think that it's important to have clear rules on which layers are allowed to talk to which other layers.
One man's quick fix for platform-xyz becomes another mans refactoring nightmare later on...

tkelman commented:
We'd actually rather remove our posix assumptions as much as possible. Ideally libuv should stand between us and either posix or win32 for as much as it can. There are plenty of places that it can't do the job completely though, and llvm doesn't abstract away the OS to the extent that would allow us to clean up codegen, debuginfo, ffi, etc. The new ORC JIT API's that are being worked on in 3.7 might be nicer in this respect, I'm not sure?

I did some quick comparisons and we have fewer Windows ifdefs than cmake or CPython, but evidently a bit more than either LLVM or luajit (or I was just searching for the wrong defines in those projects). Julia's C core is pretty small by language implementation standards, but it tends to be grouped by functionality rather than having centralized platform differences consolidated into a handful of headers or source files.

vtjnash commented:
I believe that LLVM separates a lot of it's host specific code into separate files (including autogenerated tblgen #include files), so that fewer ifdef's can cover more code.

I don't really care how many ifdef's are needed for the C code. My goal is to largely eliminate the need for them in user code inside of Julia, by providing a user interface that can be consistently implemented across platform. this effort is greatly assisted by libuv, since it provides a consistent interface for dealing with a variety of platform-specific issues (generally related to I/O).

jakebolewski commented:
Ok, rebased.  

@amitmurthy I had to change `--worker custom` to `--worker=custom` (`--worker` is the same as `--worker=default`)  reflecting recent changes to the cluster manager interface.

amitmurthy commented:
OK. Nice work on the cleanup.

JeffBezanson commented on base/client.jl:
```diff
@@ -210,120 +208,126 @@ function init_bind_addr(args::Vector{UTF8String})
+                startup = false
+            end
+            # load ~/.julia_history file
+            no_history_file = bool(opts.historyfile)
```
Should there be a `!` here?

jakebolewski replied:
I switched this to `history_file=true/false` which follows other changes in this PR.

tkelman commented on test/cmdlineargs.jl:
```diff
@@ -0,0 +1,168 @@
+    @test !success(`$exename --load`)
+
+    # --cpu-target
+    @test !success(`$exename -C invalidtarget`)
```
This fails when you run the tests with `julia-debug` on Windows.

The `cmdlineargs` test should also probably be a bit earlier in the `chosetests` list, it takes a good 30-40 seconds to run (even longer on osx travis).

tkelman replied:
cc @ihnorton the failure here with `julia-debug.exe` might be a bug actually, not sure. We should probably try to make the behavior consistent across platforms though. Does julia-debug load `sys.so`/`sys.dylib` on Linux or OSX?

jakebolewski replied:
The tests pass with `julia-debug` on OSX, but I get the same error with the debug build if I delete `sys.dylib`.  Removing the `-C target` and `--target=target` allow the tests to pass.

```julia
    * cmdlineargs         exception on 1: ERROR: LoadError: test failed: !(success(@cmd "\$exename -C invalidtarget"))
 in expression: !(success(@cmd "\$exename -C invalidtarget"))
 in error at error.jl:19
 in default_handler at test.jl:25
 in do_test at test.jl:50
 in anonymous at no file:58
 in runtests at /Users/jacobbolewski/Julia/julia/test/testdefs.jl:78
 in anonymous at multi.jl:656
 in run_work_thunk at multi.jl:617
 in remotecall_fetch at multi.jl:690
 in remotecall_fetch at multi.jl:705
 in anonymous at task.jl:1406
while loading cmdlineargs.jl, in expression starting on line 1
ERROR: LoadError: LoadError: test failed: !(success(@cmd "\$exename -C invalidtarget"))
 in expression: !(success(@cmd "\$exename -C invalidtarget"))
 in error at error.jl:19
 in default_handler at test.jl:25
 in do_test at test.jl:50
 in anonymous at no file:58
 in runtests at /Users/jacobbolewski/Julia/julia/test/testdefs.jl:78
 in anonymous at multi.jl:656
 in run_work_thunk at multi.jl:617
 in remotecall_fetch at multi.jl:690
 in remotecall_fetch at multi.jl:705
 in anonymous at task.jl:1406
while loading cmdlineargs.jl, in expression starting on line 1
while loading /Users/jacobbolewski/Julia/julia/test/runtests.jl, in expression starting on line 3
```

ihnorton replied:
> Does julia-debug load sys.so/sys.dylib on Linux or OSX?
Yes.

I don't think this is a very useful test. Running it manually I get the following message, but the REPL still starts:
```
$ usr/bin/julia-debug.exe -C invalidtarget
'invalidtarget' is not a recognized processor for this target (ignoring processor)
...
```
Whereas the logic of the test suggests that it should error out. I'm not sure where that would happen, because the only place we check `cpu_target` right now is [here](https://github.com/JuliaLang/julia/blob/ae714186f3caaed811ce4c7e18659ba57a247369/src/dump.c#L188-L222), to make sure there is no conflict with the sysimg (which in any case is not loaded on windows under `julia-debug`).

jakebolewski replied:
We could turn the test off, I don't really care.  We do fail here with the sysimg but I don't know why anyone would rely on that behavior.
